Source/BoundaryConditions/WarpX_PEC.H

```diff
@@ -295,6 +309,331 @@ using namespace amrex;
+     *        back into the simulation domain (if a reflecting particle
+     *        boundary is used), or the opposite charge density is deposited
+     *        back in the domain to capture the effect of an image charge.
+     *
+     *
+     * \param[in] dom_lo       index value of the lower domain boundary (cell-centered)
+     * \param[in] dom_hi       index value of the higher domain boundary (cell-centered)
+     * \param[in] nguard       number of guard cells along each dimension
+     * \param[in] ijk_vec      indices along the x(i), y(j), z(k) of the rho Array4
+     * \param[in] n            index of the MultiFab component being updated
+     * \param[in] rho          field data to be updated if (ijk) is at the boundary
+                               or a guard cell
+     * \param[in] is_nodal     staggering of the field data being updated.
+     * \param[in] fbndry_lo    Field boundary type at the lower boundaries
+     * \param[in] fbndry_hi    Field boundary type at the upper boundaries
+     * \param[in] pbndry_lo    Particle boundary type at the lower boundaries
+     * \param[in] pbndry_hi    Particle boundary type at the upper boundaries
+     */
+    AMREX_GPU_DEVICE AMREX_FORCE_INLINE
+    void SetRhofieldFromPEC (const amrex::IntVect & dom_lo,
+                             const amrex::IntVect & dom_hi,
+                             const amrex::IntVect & nguard,
+                             const amrex::IntVect & ijk_vec,
+                             const int n,
+                             amrex::Array4<amrex::Real> const& rho,
+                             const amrex::IntVect & is_nodal,
+                             amrex::GpuArray<int, 3> const& fbndry_lo,
+                             amrex::GpuArray<int, 3> const& fbndry_hi,
+                             amrex::GpuArray<ParticleBoundaryType, 3> const& pbndry_lo,
+                             amrex::GpuArray<ParticleBoundaryType, 3> const& pbndry_hi )
+    {
+        amrex::IntVect ijk_mirror;
+        // Loop over all dimensions
+        for (int idim = 0; idim < AMREX_SPACEDIM; ++idim) {
+            ijk_mirror = ijk_vec;
+            // Loop over sides, iside = 0 (lo), iside = 1 (hi)
+            for (int iside = 0; iside < 2; ++iside) {
+                // check if the boundary in this direction is a PEC, otherwise
+                // continue to the next direction
+                if (!( (iside == 0 )
+                    ? is_boundary_PEC(fbndry_lo, idim)
+                    : is_boundary_PEC(fbndry_hi, idim) )) continue;
+
+                // mirror guard cell outside the domain by "ig" number cells, in direction, idim
+                const int ig = ( (iside == 0)
+                                ? (ijk_vec[idim] - dom_lo[idim])
+                                : (dom_hi[idim] + is_nodal[idim] - ijk_vec[idim]) );
+
+                if (ig >= is_nodal[idim]  && ig <= nguard[idim]) {
```

Comment:
`ig` gives the mirrored guard cell number. For nodal fields, the 0'th guard cell is on the boundary, so sort of both inside and outside the domain but we don't need to do anything to it since in `SetRhoFieldOnPEC` it will be set to 0. For cell-centered fields the 0'th ghost cell is the cell center next to the PEC (outside the domain) and so it mirrors the first valid entry in the domain, and we do need to reflect it's density back into the domain. Does that clear things up?

Reply:
Thank you, it does. Maybe it could be worth adding a similar inline comment (about the differences between the nodal and cell-centered cases) in the various functions, wherever `ig` is defined or wherever you think it's most appropriate, just to make it easier to read through the code. What do you think?

Reply:
Yeah, that sounds good to me. I'll do so!